midi_tokenizers/no_loss_tokenizer.py

```diff
@@ -0,0 +1,171 @@
+            self.vocab.append(f"NOTE_OFF_{pitch}")
+
+        for vel in range(self.n_velocity_bins):
+            self.vocab.append(f"VELOCITY_{vel}")
+
+        time_vocab = self._time_vocab()
+        self.max_time_token = time_vocab[-1]  # Maximum time token
+        return self.vocab
+
+    def _time_vocab(self):
+        time_vocab = []
+        token = self.eps
+
+        # Generate time tokens with exponential distribution
+        while token < 1:
+            time_vocab.append(f"{token}s")
+            token *= 2
+        return time_vocab
+
+    def quantize_frame(self, df: pd.DataFrame):
+        df["velocity_bin"] = np.digitize(df.velocity, self.velocity_bin_edges) - 1
+        return df
+
+    def _build_velocity_decoder(self):
+        self.bin_to_velocity = []
+        for it in range(1, len(self.velocity_bin_edges)):
+            velocity = (self.velocity_bin_edges[it - 1] + self.velocity_bin_edges[it]) / 2
+            self.bin_to_velocity.append(int(velocity))
+        return self.bin_to_velocity
+
+    @staticmethod
+    def _notes_to_events(notes: pd.DataFrame):
+        """
+        Convert MIDI note dataframe into a dict with on/off events.
+        """
+        note_on_df: pd.DataFrame = notes.loc[:, ["start", "pitch", "velocity_bin"]]
+        note_off_df: pd.DataFrame = notes.loc[:, ["end", "pitch"]]
+
+        note_off_df["time"] = note_off_df["end"]
+        note_off_df["event"] = "NOTE_OFF"
+        note_on_df["time"] = note_on_df["start"]
+        note_on_df["event"] = "NOTE_ON"
+
+        note_on_events = note_on_df.to_dict(orient="records")
+        note_off_events = note_off_df.to_dict(orient="records")
+        note_events = note_on_events + note_off_events
+
+        note_events = sorted(note_events, key=lambda event: event["time"])
+        return note_events
+
+    def tokenize(self, notes: pd.DataFrame) -> list[str]:
+        notes = self.quantize_frame(notes)
+        tokens = []
+        # Time difference between current and previous events
+        previous_time = 0
+        note_events = self._notes_to_events(notes=notes)
+
+        for current_event in note_events:
+            # Check timing beginning with the largest step
+            current_step = self.max_time_token
+            # Calculate the time difference between current and previous event
+            dt = current_event["time"] - previous_time
+            filling_dt = 0  # filled time difference
+            # Fill the time gap
+            while True:
+                if filling_dt + current_step > dt:
+                    # Select time step that will fit into the gap
+                    current_step /= 2
+                else:
+                    # Fill the gap with current time token
+                    tokens.append(f"{current_step}s")
+                    filling_dt += current_step
+
+                if dt - filling_dt < self.eps:
+                    # Exit the loop when the gap is filled
+                    break
```

Comment:
Can we move this to a separate method and add a short docstring explaining what it does? The usage should look something like this:

```python
time_tokens = self.tokenize_time_distance(dt)
tokens += time_tokens
```